cEP-0018.md

```diff
@@ -0,0 +1,164 @@
+5. `antlrastlib` will have an AST loader class that will be responsible
+   for loading the AST of a given file depending on the extension.
+   For e.g, a .c file will be loaded using the parser from
+   `antlrastlib` for c language.
+6. Another AST walker class will be responsible for providing endpoints for
+   traversing the AST using methods such as `get_next_node` and
+   `get_previous_node`.
+
+## Management of the new `coala-ast` repository
+
+Managing a new repository is a heavy task, and this will be highly automated as
+follows:
+
+1. The parsers in coala-antlr-ast/antlrparsers would be generated and
+   pushed via travis-builds whenever a new commit is pushed.
+2. The cib tool can be enhanced to deal with the installation of bears that
+   require only some specified parsers (for e.g a PyUnusedVarBear would
+   only require parser for python)
+3. The cib tool can also trigger specialised builds and download the newly
+   generated parser on the fly
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations within antlrastlib:
+
+```python
+import antlrparsers
+import antlr4
+
+class ASTLoader():
+    lexerMapping = {
+        'c' : antlrparsers.clexer,
+        'py' : antlrparsers.pylexer,
+        'js' : antlrparsers.jslexer,
+        ...
+    }
+    parserMapping = {
+        'c' : antlrparsers.cparser,
+        'py' : antlrparsers.pyparser,
+        'js' : antlrparsers.jsparser,
+        ...
+    }
+
+    @staticmethod
+    def loadFile(file,filename):
+        """ Loads file's AST into memory """
+        ext = get_extension(filename)
+        if ext is None:
+            raise FileExtensionNotSupported
+        inputStream = file.readlines()
+        lexer = lexerMapping[ext](inputStream)
+        parser = parserMapping[ext](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """ Modify the internal tree variable and return the next node """
+
+    def get_prev_node(self):
+        """ Modify the internal tree variable and return prev node """
+
+    def reset(self):
+        """ Reset the internal tree variable to tree root """
+
+    def isTerminalNode(node):
+        """ Determine if the node is terminal """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coalib.bears.LocalBear import Bear
+from antlrastlib import ASTWalker
+
+class ASTBear(LocalBear):
+    walker = None
+
+    def initialise(file, filename):
+        self.walker = ASTWalker(file, filename)
+
+    def run(self,
+            filename,
+            file,
+            tree,
+            *args,
+            dependency_results = None,
+            **kwargs
+            ):
+        raise NotImplementedError           # Needs to be done by bear
```

Comment:
ahem. https://github.com/coala/cEPs/pull/124#discussion_r187032788

Reply:
Ah
This form of error raising is deprecated
I thought it was about spaces besides the operators :sweat_smile: 
Will fix it :D